deepchem/data/data_loader.py

```diff
@@ -156,32 +141,46 @@ def featurize_mol_df(df, featurizer, field, verbose=True, log_every_N=1000):
 
 
 class DataLoader(object):
+  """Handles loading/featurizing of data from disk.
+
+  The `Featurizer` objects can featurize provided input into
+  numpy arrays but doesn't generate `Dataset` objects. You can
+  of course wrap numpy arrays into `Dataset` objects with
+  `dc.data.NumpyDataset`, but you might face some difficulty
+  with larger dataset processing. The main use of `DataLoader`
+  and its child classes is to make it easier to load large
+  datasets into `Dataset` objects.` You won't ever "need" to use
+  a `DataLoader` but might often find it convenient when
+  processing larger datasets.
```

Comment:
This description is kind of confusing.  A DataLoader does or doesn't create Dataset objects?  It does or doesn't have problems with large datasets?  This seems to imply both.

If DataLoader shouldn't ever be used directly, only subclasses, would it make sense for the constructor to check `self.__class__ is DataLoader` and throw an exception?

Reply:
I've rewritten the docstring so it's hopefully more useful. I've also added a constructor check which prevents direct instantiation of `DataLoader`

Reply:
Much better!  The one thing that still isn't explicitly stated anywhere is that `featurize()` returns a Dataset.  Maybe update the docstring for it to state explicitly that it creates and returns a DiskDataset?

Would it be even clearer if the method were called `create_dataset()`?

Reply:
I think `create_dataset()` would be a better name. `featurize()` is unfortunate since it causes confusion with `Featurizer.featurize()`. It's likely that `DataLoader.featurize()` is used in a lot of downstream code though (all the moleculenet loaders for example). Perhaps we can keep `featurize()` around as a wrapper with a deprecation flag and have it call `create_dataset()` under the hood. We can phase out `DataLoader.featurize()` over a couple of releases

Reply:
Sounds perfect.